### ISSUE TITLE
Choosing an error function

### ISSUE THREAD

**Issue by stefanpeidli** [stefanpeidli/GoNet#8]
From C. Bishop Neuronal Networks for Pattern Recognition, pages: 195ff

> We have derived the sum-of-squares  error function  from  the principle of maxi-
mum  likelihood  on the assumption  of Gaussian  distributed  target  data.  Of course 
the  use of a sum-of-squares  error  does not  require the target  data  to have a Gaus-
sian distribution. Later in this chapter we shall consider the least-squares  solution 
for  an  example  problem  with a strongly  non-Gaussian  distribution.  However, as 
we shall  see, if we use a sum-of-squares  error,  then  the  results  we obtain  cannot 
distinguish  between  the  true  distribution  and  any  other  distribution  having  the 
same  mean  and  variance. 

In our case this means: 
If for example one particular move is played super often and succesfully, our distribution has more similarities to a Dirac-Distribution than a Gaussian distribution. Thus we get bad results in Absolute Error.

A solution would be:
We should choose the error function with respect to the expected output distribution.

**Comment by stefanpeidli** [comment]
The "error" between two probability distributions can be quantified by a **Statistical distance**. The idea is to think of a multivariate probabilistic measure space, equipped with a (quasi) metric.
On wikipedia are some examples of [statistical distances](https://en.wikipedia.org/wiki/Statistical_distance#Examples).

**Comment by stefanpeidli** [comment]
The **Kullback-Leibler divergence** as error function gives ultra good results.

**Comment by stefanpeidli** [comment]
The Kullback-Leibler divergence has some downsides:
- The K-L div does **not** punish having non-zero values in the suggested distribution for a zero-valued target. This might be a big problem, because then our Network only learns what is a good move, but not what a bad move is.

Not really problems, still worrying about:
- It tends to be numerically unstable (partially fixed).
- It is no metric in a discrete probability measure space. But that would be quite neat.
- (Convexity holds only for pairs of distributions. Thus changing only one part won't result in a convex Error surface, which is (is it?) needed for Gradient Descent to deliver optimality.) This actually is not a problem according to Bernhard.

Ideas for Solutions:
- Modify the K-L div to our needs
- Search for different approaches
- Model an error function "from scratch"

**Comment by paddyooo** [comment]
what about Relu?

**Comment by stefanpeidli** [comment]
First:
Relu is an **activation function**. This issue is about **error functions**. Those two things are totally different.

From Wikipedia:

- Second:
Dying ReLU problem: ReLU neurons can sometimes be pushed into states in which they become inactive for essentially all inputs. In this state, no gradients flow backward through the neuron, and so the neuron becomes stuck in a perpetually inactive state and "dies." In some cases, large numbers of neurons in a network can become stuck in dead states, effectively decreasing the model capacity. This problem typically arises when the learning rate is set too high. It may be mitigated by using Leaky ReLUs instead.

- Third:
Non-zero centered.
Zero centered is needed for our weight initialization method.

- Forth:
Sparse activation: For example, in a randomly initialized network, only about 50% of hidden units are activated (having a non-zero output).

Thus I think Relus won't be relevant for us.

**Comment by paddyooo** [comment]
or ELU https://www.quora.com/How-does-ELU-activation-function-help-convergence-and-whats-its-advantages-over-ReLU-or-sigmoid-or-tanh-function

**Comment by paddyooo** [comment]
ah ok, our activation function was tanh right?

**Comment by stefanpeidli** [comment]
Yes.
You can find the issue for activation function among the [closed issues](https://github.com/stefanpeidli/GoNet/issues?q=is%3Aissue+is%3Aclosed).

**Comment by stefanpeidli** [comment]
We could also try using the [Hellinger distance](https://en.wikipedia.org/wiki/Hellinger_distance) as error function.

It is closely related to the K-L divergence but with the following advantages:

It is indeed a metric:
> The Hellinger distance forms a bounded metric on the space of probability distributions over a given probability space.

It is bounded by 1, allowing easier interpretation of the error value:
> The maximum distance 1 is achieved when P assigns probability zero to every set to which Q assigns a positive probability, and vice versa.

**Comment by stefanpeidli** [comment]
The Hellinger distance seems to give a more steady learning.
A first simple test with boards from only a single game yields the followig error curve:

![image](https://user-images.githubusercontent.com/24811160/33728895-4b1f3f9c-db7c-11e7-85e9-e7115882343f.png)

This seems promissing enough to keep testing this approach.

**Comment by stefanpeidli** [comment]
I have found a neat site on [common error functions](https://stats.stackexchange.com/questions/154879/a-list-of-cost-functions-used-in-neural-networks-alongside-applications). I will try to implement most of its mentioned functions.

**Comment by stefanpeidli** [comment]
I wrote a test script for comparing error functions. MSE seems to be the only known to us error function capable of distinguishing for example [0, 0, 0.2, 0.8, 0] (disturbed dirac) and [0.2, 0.2, 0.2, 0.2, 0.2] (uniform) with given target [0, 0, 1, 0, 0] (dirac). **This agains speaks for MSE instead of KLD**.

Also we came up with the idea of using affine/convex combinations of error functions.
